protocol.md

```diff
@@ -870,6 +902,12 @@ export interface TextDocumentClientCapabilities {
 ```typescript
 interface ClientCapabilities {
 	/**
+	 * The client supports receiving the result solely through $/partialResult notifications for requests from the client to the server.
+	 * If true, the server should send all data through $/partialResult and send null as the result in the final response.
```

Comment:
> the client's protocol layer can support servers that don't support streaming by transforming the full responses to a single replace op, so the method handlers would always get a stream and don't have to differentiate.

We are talking about two different clients. When I said client, I mean the consumer of the LSP wire format (i.e. vscode-languageclient). vscode-languageclient will have to support both streaming and non-streaming. Higher level APIs can be designed however is most convenient regardless of wire protocol.

> What do we gain from putting another layer on top of the jsonrpc layer

Let's table this proposal for now, I think we have a bigger problem to discuss...

> It honestly is starting to feel like the correct place for this to exist is in the jsonrpc2 spec, but that seems less practical.

Hmm, yeah this does seem like a problem. JSON RPC doesn't mention or define any sort of ordering guarantees between different requests or notifications. Even if all requests are streaming, this is a critical problem for the ordering of notifications (i.e. $/partialResult).

I see a few paths forward:
1. We implement some ordered delivery logic on top of JSON RPC. Something like this:
    - $/partialResult needs to be a request (not a notification).
    - Server has to wait for acknowledgement of previous $/partialResult before sending a new one.
    - Server has to wait for acknowledgement of previous $/partialResult before sending the final response to the original request.
2. Use timing hacks to try to mitigate out of order scenarios. We should write this into the spec.
    - Server SHOULD NOT send a $/partialResult within Xms of a previous $/partialResult for the same request.
    - Server SHOULD delay the final response until Xms after the previous $/partialResult for the same request.
3. Treat $/partialResult as "best effort" and out of order.
    - Require the server to send the complete result in the final response.
    - One of these
        - Can't use JSON PATCH as the protocol since $/partialResults are unordered, or restrict usage of JSON PATCH to operations that still work out of order (e.g. append to array, only set a give key once). This would prevent us from sending data and decorating it later.

Of these, I favor (1). (2) feels too hacky for a spec and (3) seems like more work than (1) and less optimal than (1).